File: ducklin-core/src/main/java/io/github/ducklin/migration/items/groups/FlexItemGroup.java
package io.github.ducklin.migration.items.groups;

import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import io.github.ducklin.migration.guide.SlimefunGuideMode;
import io.github.ducklin.migration.items.SlimefunItem;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import io.github.ducklin.migration.items.ItemGroup;
import io.github.ducklin.migration.player.PlayerProfile;


public abstract class FlexItemGroup extends ItemGroup {

    @ParametersAreNonnullByDefault
    protected FlexItemGroup(NamespacedKey key, ItemStack item) {
        this(key, item, 3);
    }

    @ParametersAreNonnullByDefault
    protected FlexItemGroup(NamespacedKey key, ItemStack item, int tier) {
        super(key, item, tier);
    }

    @Override
    public final boolean isVisible(@Nonnull Player p) {
        /*
         * We can stop this method right here.
         * We provide a custom method with more parameters for this.
         * See isVisible(...)
         */
        return true;
    }


    @ParametersAreNonnullByDefault
    public abstract boolean isVisible(Player p, PlayerProfile profile, SlimefunGuideMode layout);

    public abstract void open(Player p, PlayerProfile profile, SlimefunGuideMode layout);

    @Override
    public final void add(@Nonnull SlimefunItem item) {
        throw new UnsupportedOperationException("You cannot add items to a FlexItemGroup!");
    }

    @Override
    public final @Nonnull List<SlimefunItem> getItems() {
        throw new UnsupportedOperationException("A FlexItemGroup has no items!");
    }

    @Override
    public final boolean contains(SlimefunItem item) {
        throw new UnsupportedOperationException("A FlexItemGroup has no items!");
    }

    @Override
    public final void remove(@Nonnull SlimefunItem item) {
        throw new UnsupportedOperationException("A FlexItemGroup has no items, so there is nothing remove!");
    }

}

File: ducklin-api/src/main/kotlin/io/github/ducklin/api/items/groups/FlexItemGroup.kt
package io.github.ducklin.api.items.groups

import io.github.ducklin.api.guide.DucklinGuideMode
import io.github.ducklin.api.items.DucklinItem
import io.github.ducklin.api.items.ItemGroup
import io.github.ducklin.api.player.PlayerProfile
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

abstract class FlexItemGroup (
    key: NamespacedKey,
    item: ItemStack,
    tier: Int = 3
) : ItemGroup(key, item, tier) {

    override fun isVisible(p: Player): Boolean = true

    abstract fun isVisible(p: Player, profile: PlayerProfile, layout: DucklinGuideMode): Boolean
    abstract fun open(p: Player, profile: PlayerProfile, layout: DucklinGuideMode)

    fun getItems(): List<DucklinItem> = throw UnsupportedOperationException("You cannot add items to a FlexItemGroup!")

    override fun add(item: DucklinItem) = throw UnsupportedOperationException("You cannot add items to a FlexItemGroup!")

    override fun contains(item: DucklinItem?): Boolean = throw UnsupportedOperationException("A FlexItemGroup has no items!")

    override fun remove(item: DucklinItem) = throw UnsupportedOperationException("A FlexItemGroup has no items, so there is nothing remove!")

}